### PR TITLE
prefer active health check during outlier ejection wait

### PR DESCRIPTION
**Pull request by ramaraochavali**
Commit Message:prefer active health check during outlier ejection wait
Additional Description:
Risk Level:Low
Testing:Added
Docs Changes:
Release Notes: Updated

Fixes https://github.com/envoyproxy/envoy/issues/24217

### CODE REVIEW

**Review by snowp — CHANGES_REQUESTED** [review]
This seems right to me but is there any way to test this interacting with actual outlier detection? For example, are we going to hit any asserts in that code due to the flag being cleared outside of the outlier detection code?

File: source/common/upstream/health_checker_base_impl.cc
Comment (ramaraochavali) [reply source/common/upstream/health_checker_base_impl.cc:299]: Sure. Will change

**Comment by ramaraochavali** [comment]
>  but is there any way to test this interacting with actual outlier detection?

I validated the code and did not find any problem. However, if you have a test in mind let me know - I can add. Otherwise I will take a look at outlier detection tests and see if it is possible to add there.

**Comment by snowp** [comment]
I think perhaps just an outlier detection test where you manually flip the flag out of band with outlier detection might be sufficient? I think that would simulate what happens if something else resets that flag

/wait-any

File: source/common/upstream/outlier_detection_impl.cc
Comment (snowp) [inline source/common/upstream/outlier_detection_impl.cc:316]: Seems like a consequence of this is that not only do we uneject the host on hc success, we also reset any outlier detector counters etc on a successful hc (when the host was previously unhealthy)? This seems in line with considering active hc as an authoritative check, but seems like a bigger change than I first anticipated. We probably want this behavior documented in the outlier detention docs and reflected in the change log

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks, a few more comments

File: test/common/upstream/outlier_detection_impl_test.cc
Comment (ramaraochavali) [reply test/common/upstream/outlier_detection_impl_test.cc:389]: No specific reason. Can remove it

File: source/common/upstream/outlier_detection_impl.cc
Comment (ramaraochavali) [reply source/common/upstream/outlier_detection_impl.cc:316]: Makes sense. Updated

**Comment by ramaraochavali** [comment]
@snowp Can you PTAL when you get chance?

**Comment by ramaraochavali** [comment]
@snowp PTAL when you get chance

File: source/common/upstream/outlier_detection_impl.cc
Comment (snowp) [inline source/common/upstream/outlier_detection_impl.cc:316]: Will this cause a regular transition from healthy -> unhealthy via active hc cause the outlier detector to count this as an unejection causing the stats to get messed up? Should this only respond to health transitions that clear the outlier check flag?

In general this approach of indirectly responding to a health change via the hc callback seems a bit brittle, though I'm not sure what the better approach would be. Maybe @mattklein123 has any ideas?

Comment (ramaraochavali) [reply source/common/upstream/outlier_detection_impl.cc:316]: > Will this cause a regular transition from healthy -> unhealthy via active hc cause the outlier detector to count this as an unejection causing the stats to get messed up? Should this only respond to health transitions that clear the outlier check flag?

I do not think. But we can restrict when only respond to health transitions that clear the outlier check flag to be 100 % safe. I will change it like that.

> In general this approach of indirectly responding to a health change via the hc callback seems a bit brittle, though I'm not sure what the better approach would be. Maybe @mattklein123 has any ideas?

I could not think of any better approach. Let us see if @mattklein123 has any ideas

**Comment by ramaraochavali** [comment]
@snowp PTAL when you get chance

File: source/common/upstream/outlier_detection_impl.cc
Comment (snowp) [inline source/common/upstream/outlier_detection_impl.cc:316]: I don't think we want to gate on this? Since the health checker doesn't care about FAILED_OUTLIER_CHECK it would only set HealthTransition::Changed if we were already failing active hc and are now succeeding

I guess `!host->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC) && host->healthFlagGet(Host::HealthFlag::FAILED_OUTLIER_CHECK)`  would be sufficient to say "we got a hc indicating that active hc is good, but the outlier detector has flagged this host so we should clear that out"? 

I think ideally we'd have an integration test operating between the active hc and the outlier detector, seems a bit tricky to make sure that things are working as expected just through unit tests

Comment (ramaraochavali) [reply source/common/upstream/outlier_detection_impl.cc:316]: >  I guess !host->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC) && host->healthFlagGet(Host::HealthFlag::FAILED_OUTLIER_CHECK) would be sufficient to say "we got a hc indicating that active hc is good, but the outlier detector has flagged this host so we should clear that out"?

But we want this behaviour of whether active health check success should clear of pending outlier detection failure should be guarded by run time flag in case some one has some issues with it. But we are defaulting it to true.

**Comment by KBaichoo** [comment]
friendly ping @snowp 

Think this is waiting on your input

File: source/common/upstream/outlier_detection_impl.cc
Comment (snowp) [reply source/common/upstream/outlier_detection_impl.cc:316]: Sorry yeah this should be gated by a feature flag, I was thinking about the `HealthTransition::Changed` check. I would think that we want to respond to every health check result from the active health checker and reset any ejected hosts regardless of whether we transitioned the active health of the host?

Comment (ramaraochavali) [reply source/common/upstream/outlier_detection_impl.cc:316]: I see what you mean. Yes. That can be removed

**Comment by ramaraochavali** [comment]
@snowp PTAL when you get chance

File: test/common/upstream/outlier_detection_impl_test.cc
Comment (snowp) [inline test/common/upstream/outlier_detection_impl_test.cc:401]: Can you do a test case where you use other transition values?

**Review by snowp — CHANGES_REQUESTED** [review]
Sorry for the slow turnaround here, just one thing

File: test/common/upstream/outlier_detection_impl_test.cc
Comment (ramaraochavali) [reply test/common/upstream/outlier_detection_impl_test.cc:429]: Done

**Review by snowp — APPROVED** [review]
LGTM, thanks!

@mattklein123 mind doing the second review?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM but we should wait to merge this till after the release (today) due to risk. Once the release is cut can you merge main which will require a release note fix? Thank you.

/wait

**Review by mattklein123 — APPROVED** [review]

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Sorry I think you have a merge issue on the release notes.

/wait

**Comment by ramaraochavali** [comment]
Ah. Sorry. Did n't realize that. Fixed now.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
